### PR TITLE
[Spot] add order placement logic

### PR DESCRIPTION
**Pull request by codchen**
- For a buy, it checks if there is enough fund (price * quantity) locked in the contract, and only place order if so.
- For a sell, it checks if there is enough quantity locked in the contract, and only place order if so.

### CODE REVIEW

File: src/balance.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+use crate::state::Balance;
+use cosmwasm_std::{Decimal, Storage};
+use cw_storage_plus::Map;
+
+// addr, denom -> balance
+pub const BALANCE: Map<(String, String), Balance> = Map::new("bal");
+
+pub fn save_balance(storage: &mut dyn Storage, account: String, symbol: String, balance: &Balance) {
```
Comment (LCyson) [inline src/balance.rs:8]: Can we add test for this file

File: src/order.rs [hunk 2]
```diff
@@ -0,0 +1,18 @@
+use crate::state::Order as OrderState;
+use cosmwasm_std::Storage;
+
+use cw_storage_plus::Map;
+
+pub const ORDER: Map<u64, OrderState> = Map::new("order");
+pub const ORDER_ACCOUNT_INDEX: Map<(String, u64), OrderState> = Map::new("oai");
+
+pub fn save_order(storage: &mut dyn Storage, order: &OrderState) {
```
Comment (LCyson) [inline src/order.rs:9]: ditto: small tests

File: src/state.rs [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    pub price: Decimal,
+    pub quantity: Decimal,
+    pub order_type: i32,
+    pub position_direction: i32,
+    pub data: String,
+}
+
+impl OrderPlacement {
+    pub fn to_order(&self) -> Result<Order, ContractError> {
+        let order_data: OrderData = match serde_json_wasm::from_str(&self.data) {
```
Comment (LCyson) [inline src/state.rs:56]: ditto quick test here

File: src/contract.rs [hunk 4]
```diff
@@ -67,12 +79,75 @@ fn process_settlements(
...
+    }
+    for order_placement in orders {
+        let order = order_placement.to_order()?;
+        let denom = if order.direction == PositionDirection::Long {
+            order.price_denom.to_owned()
+        } else {
+            order.asset_denom.to_owned()
+        };
+        let nominal = if order.direction == PositionDirection::Long {
+            order.price * order.quantity
```
Comment (LCyson) [inline src/contract.rs:114]: Can we add a comment/example of the direction definition (it's always SOL/USDC, price_denom=sol, asset_denom=usdc, Long be buying sol and short be selling sol here right)

**Review by LCyson — APPROVED** [review]

File: src/balance.rs (hunk 1 above)
Comment (codchen) [reply src/balance.rs:8]: added

File: src/order.rs (hunk 2 above)
Comment (codchen) [reply src/order.rs:9]: added

File: src/state.rs (hunk 3 above)
Comment (codchen) [reply src/state.rs:56]: added

File: src/contract.rs (hunk 4 above)
Comment (codchen) [reply src/contract.rs:114]: added

**Pull request merged**